File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:

        n=len(nums)
        ans = 0
        if n==0:
            return 0
        if n==1:
            return nums[0]

        temp1 = max(nums[0],nums[1])
        temp2 = nums[0]
        

        for i in range(2,len(nums)):
            curr = max(nums[i]+temp2, temp1)
            temp2 = temp1
            temp1 = curr

        return temp1
            


        # def dp_rob(n):
        #     if n<0:
        #         return 0
            
        #     if n==0:
        #         return nums[n]
            
        #     if dp_table[n]!=-1:
        #         return dp_table[n]
            
        #     pick =nums[n] + dp_rob(n-2)

        #     not_pick = dp_rob(n-1)

        #     return max(pick, not_pick)
        
        # return dp_rob(n-1)



        

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int size=nums.size();

        //base case
        if(size==1){
            return nums[0];
        }
        //array for max loot at each position
        int max_loot[size];

        //case for two elements in array;
        max_loot[0]=nums[0];
        max_loot[1]=max(nums[0],nums[1]);


        //case for more than two elements
        for(int i=2; i<size; i++){
            max_loot[i]=max(max_loot[i-2]+nums[i], max_loot[i-1]);
        }

        return max_loot[size-1];
    }
};